### ISSUE TITLE
Add Priority Queue/Heap Data structure to standard library

### ISSUE THREAD

**Issue by alexsnezhko3** [grain-lang/grain#1383]
[Priority queues](https://en.wikipedia.org/wiki/Priority_queue) are data structures that internally attach a priority value to each element in them. Among other applications, they allow for processing a collection of elements in the order of highest/lowest priority first as needed, and are also useful for sorting elements (see [heap sort](https://en.wikipedia.org/wiki/Heapsort)). I believe that priority queues are a fairly useful data structure, and may be a good idea to include directly in the standard library

Proposal:
- Create a new standard library module named "priorityqueue" (or maybe an alternate casing such as "priority-queue" or "priorityQueue", so far all of the stdlib modules are a single word so I don't know which casing should be used going forward for multi-word module names)
- The priority queue will be a mutable data structure, and will be implemented as a [binary heap](https://en.wikipedia.org/wiki/Binary_heap) with internal storage of elements in an array
- API:
  - `make(comparatorFn: (a, a) -> Bool) -> PriorityQueue<a>`, taking a comparator function determining how priority should be assigned comparatively to elements based on their "objective" value ordering. For example, with a heap of numbers, returning `(firstArg < secondArg)` would create a min-heap, where the smallest elements will be pulled first when polling the priority queue.
  - `makeSized(size: Number, comparatorFn: (a, a) -> Bool) -> PriorityQueue<a>`: like `make` but allows for defining an initial size for the internal array for performance reasons if desired.
  - `size(pq: PriorityQueue<a>) -> Number`: returns the number of elements in the priority queue.
  - `isEmpty(pq: PriorityQueue<a>) -> Bool`: returns whether or not the priority queue has elements in it.
  - `push(val: a, pq: PriorityQueue<a>) -> Void`: add a new element to the priority queue, inserting it in the appropriate priority order.
  - `pop(pq: PriorityQueue<a>) -> Option<a>`: removes and returns the "highest priority" element in the queue in a Some variant, or None if there are no more elements in the queue.
  - `peek(pq: PriorityQueue<a>) -> Option<a>`: returns the "highest priority" element in the queue in a Some variant, or None if there are no more elements in the queue.
  - `drain(pq: PriorityQueue<a>) -> List<a>`: empties the priority queue, returning a list of all of the elements removing.
  - `fromArray(arr: Array<a>) -> Void`: creates a new priority queue, loading all of the elements from the array into the queue.

**Comment by ospencer** [comment]
This looks great!

Here are a few resolutions/slight adjustments, and you can get started on this!

- use `priorityqueue` as the casing (we do this for `bigint` as well)
- mutable data structure is fine
- `make` and `makeSized` should probably use a comparator from the standard library to match our other APIs. This doesn't currently exist, but I've been meaning to write one. Not against a version of `make` that lets you supply your own comparator, but I don't have a great name for what we'd call it. I do sort of feel like we've been holding out on this until Grain has optional arguments 😅. You can write that version first, and we'll implement a version of`Pervasives.compare` to use for the regular `make` and `makeSized`. cc @phated for help with naming the user-supplied comparator version.
- I'd prefer `fromList` rather than `fromArray`, but it's fine to have both.

Otherwise the API looks great!

cc @peblair

**Comment by peblair** [comment]
@ospencer, do you think we should maybe have both an immutable and a mutable version of this? Some users may prefer immutability.

**Comment by ospencer** [comment]
Yeah, ideally we have mutable/immutable versions of all of the data structures. The idea is that we add `Immutable` to the start of the structure name; i.e. `ImmutableMap`, `ImmutableStack`, `ImmutableQueue`, etc.

In the interest of practicality, we've started with all of the mutable ones.

I suppose we should make issues for the immutable versions since people would probably volunteer to implement them 😛

So tl;dr yes we'd like an immutable version, but we don't need it immediately.

**Comment by alexsnezhko3** [comment]
Sounds good; I have a reference implementation for a mutable binary heap handy from a data structures class I had a few years ago :P . I'm not too familiar with efficient immutable data structures, but I found a [blog post](https://andrew.gibiansky.com/blog/haskell/finger-trees/) I'll try to play around with for an immutable implementation.

As far as having mutable and immutable versions of data structures in the standard library, would the naming follow that the immutable versions have "Immutable" in the name whereas the mutable versions are just the name? e.g. `"immutablepriorityqueue"` and `"priorityqueue"`?

**Comment by ospencer** [comment]
Yup, that's exactly right.

**Comment by alexsnezhko3** [comment]
For the sake of making changes easier once the comparator gets added to the standard library, what will the comparator look like? WIll it be like Java and other langs where a negative return value means a is before b, a positive means b is before a, and 0 means they are equal in ranking? Or will it be simply a true/false returning one like I outlined?

**Comment by alexsnezhko3** [comment]
@phated does `makeWithComparator`, `makeSizedWithComparator`, and `fromListWithComparator` sound good for the naming with custom comparators? I see that there are a few examples of functions in the standard library that have "with" in their name

**Comment by phated** [comment]
@alexsnezhko3 I don't like comparator functions at all 😛 As you probably noticed with Map and Set, they don't use any. Comparators will be supported when we can specialize modules (or when we have optional arguments).

**Comment by ospencer** [comment]
Comparators are particularly important for priority queues as that's how you define priority, so we have to have them. `Pervasives.compare` will work for some simple use cases but most of the time you'll need to write your own for the priority queue to be useful.

**Comment by phated** [comment]
> most of the time you'll need to write your own for the priority queue to be useful.

Then the `make` should just take a comparator. Why are we trying to explode the # of API functions? Just make everything that needs a comparator take a comparator and if you want to use `compare`, you'll just pass that.

**Comment by phated** [comment]
That being said, it seems like the comparator function should return an enum instead of a boolean.

**Comment by ospencer** [comment]
If we had optional arguments it'd be a non-issue, but while we don't it usually ends up resulting in more functions. Like you, I really value consistency in the standard library, and I wouldn't want `make` to have a signature that's far from the signature of `Map.make` or `Set.make`.

If we really don't want multiple functions and we just want one that takes a comparator, then I wouldn't call this version `make`. It'd need to be `makeWithComparator` or whatever other name we choose.

As for the result of the comparator function, that's something we could explore in a breaking release of Grain, but we already have functions in the standard library that work on comparators of the type `(a, a) -> Number`.

**Comment by phated** [comment]
In my mind, consistency isn't "this function signature needs to be exactly the same", especially when the data structure **does not make sense** without the comparator. To me, `make` is the grain equivalent of `new Class()` and those constructors take the things that are actually need to make the data structure work. For comparison, `Array.make` takes a length and value to add to each element.

**Comment by phated** [comment]
> As for the result of the comparator function, that's something we could explore in a breaking release of Grain, but we already have functions in the standard library that work on comparators of the type (a, a) -> Number.

Yeah, then we need to use `Number` for now but we should open an issue to revisit with an enum because number is super confusing in JS.

**Comment by ospencer** [comment]
I generally agree with that, though in this case I don't think it's that different from Map and Set. Map and Set aren't ordered, but they still use a default comparator for `make()`. A plain `make()` for a priority queue makes sense to me if you're fine with Grain's natural ordering. If we wanted different comparisons for Map and Set before we got optional arguments, it'd be in the form of `makeWithComparator`.

**Comment by ospencer** [comment]
I'm not super hung up on it, so if you think just making `make` take a comparator is the way to go, that's fine.

**Comment by alexsnezhko3** [comment]
Ok, I guess I'll just go with `make` then? I can see the appeal of both though. One other case to be made for `make` with comparator is to be consistent with `sort` on list/array, which currently does not have a comparator-less API

**Comment by ospencer** [comment]
Yeah, let's go with that! Easy enough to update if we change our minds 🙂
